feat: add githubgen tool

mowies opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36691

atoulme commented on githubgen/codeowners.go:
```diff
@@ -0,0 +1,231 @@
+const allowlistHeader = `# Code generated by githubgen. DO NOT EDIT.
+#####################################################
+#
+# List of components in OpenTelemetry Collector Contrib
```
hardcoded to contrib

atoulme commented on githubgen/codeowners.go:
```diff
@@ -0,0 +1,231 @@
+const codeownersHeader = `# Code generated by githubgen. DO NOT EDIT.
+#####################################################
+#
+# List of codeowners for OpenTelemetry Collector Contrib
```
hardcoded, we might need to make that configurable.

atoulme commented on githubgen/codeowners.go:
```diff
@@ -0,0 +1,231 @@
+# https://help.github.com/en/articles/about-code-owners
+#
+
+* @open-telemetry/collector-contrib-approvers
```
same

atoulme commented on githubgen/codeowners.go:
```diff
@@ -0,0 +1,231 @@
+const distributionCodeownersHeader = `
+#####################################################
+#
+# List of distribution maintainers for OpenTelemetry Collector Contrib
```
also hardcoded

atoulme commented on githubgen/codeowners.go:
```diff
@@ -0,0 +1,231 @@
+		for stability := range m.Status.Stability {
+			if stability == unmaintainedStatus {
+				unmaintainedList += key + "/\n"
+				unmaintainedCodeowners += fmt.Sprintf("%s/%s @open-telemetry/collector-contrib-approvers \n", key, strings.Repeat(" ", data.maxLength-len(key)))
```
hardcoded as well here.

atoulme commented on githubgen/codeowners.go:
```diff
@@ -0,0 +1,231 @@
+				owners += " "
+				owners += "@" + owner
+			}
+			codeowners += fmt.Sprintf("%s/%s @open-telemetry/collector-contrib-approvers%s\n", key, strings.Repeat(" ", data.maxLength-len(key)), owners)
```
hardcoded as well here

mowies commented:
@atoulme thanks for the review. I will add at least a sanity check unit test in this PR which will already need some slight refactoring and then do any further changes in follow-ups.

mowies commented:
@atoulme this is basically ready but i can't get `moq` to work on windows for some reason 🤦🏼

mx-psi commented:
@pjanotti If you have the time, could you take a quick look at the failing CI check and try to help us figure out why it is failing?

pjanotti commented:
@mowies the issue can be solved by creating a `moq.exe` at the `.tools` directory. When in the make shell env the `bash.exe` takes care of that, but, I suspect that `go generate` (as it should) uses the Windows API. A quick workaround will be to `cp .\.tools\moq .\.tools\moq.exe`

mowies commented:
@pjanotti wow thanks for the help, I don't think I would have found that by myself at all 😅 
seems like everything works now 🙌🏼

dmathieu commented on githubgen/issuetemplates.go:
```diff
@@ -0,0 +1,70 @@
+	switch path[0] {
+	case "receiver", "exporter", "extension", "processor", "connector":
+		path[1] = strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(path[1], "internal"), "extension"), "exporter"), "connector"), "processor"), "receiver")
+		path[len(path)-1] = strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(strings.TrimSuffix(path[len(path)-1], "internal"), "extension"), "exporter"), "connector"), "processor"), "receiver")
```
This looks like it shouldn't be hardcoded, but configurable per repository?

mowies commented:
@atoulme @dmathieu thanks for the reviews! I will implement the feedback along with more unit tests in a follow up PR so that all that gets easier to review.

mx-psi reviewed (approved):
Approving based on the comment above

mowies commented:
follow up PR that addresses all the feedback here and more is ready for review: https://github.com/open-telemetry/opentelemetry-go-build-tools/pull/655
